### ISSUE TITLE
Segmentation fault when loading capture

### ISSUE THREAD

**Issue by bobtedbob** [baldurk/renderdoc#2268]
## Description

I have a very simple app drawing two quads. When running the application outside of renderdoc I get a validation error (described in the next seciton) but the application still runs. The issues happens when trying to open the first capture in renderdoc and it is here that a segmentation fault occurs.

## Steps to reproduce

The below steps show the setup to reproduce the validation error based on my code base (which I don't want to share) : 

I have the following extension enabled

VK_STRUCTURE_TYPE_VALIDATION_FEATURES_EXT

And I enable this feature :

VK_VALIDATION_FEATURE_ENABLE_SYNCHRONIZATION_VALIDATION_EXT

And I get this error when running (one per buffer used)

[VULKAN DEBUG CALLBACK][ERROR][VALIDATION] Validation Error: [ SYNC-HAZARD-READ_AFTER_WRITE ] Object 0: handle = 0x710000000071, type = VK_OBJECT_TYPE_BUFFER; | MessageID = 0x4cb75b2d | vkCmdDrawIndexed: Hazard READ_AFTER_WRITE for index VkBuffer 0x710000000071 in VkCommandBuffer 0x7ff59ce903d0[]. Access info (usage: SYNC_VERTEX_INPUT_INDEX_READ, prior_usage: SYNC_TRANSFER_TRANSFER_WRITE, write_barriers: 0, command: vkCmdCopyBuffer, seq_no: 9, reset_no: 1).

The below is how to reproduce the segmentation fault in renderdoc itself : 

When loading this simple app up in renderdoc and loading the frame 0 capture I get a segmentation fault and the application crashes to the command line.

Some times the application crashes when I load the capture, other times when I expand the events in the capture.

Loading the extension is not the cause of the segmentation fault but the error might help with what is causing the crash. When I load the frame 1 capture (which doesn't have the vkCmdCopyBuffer run) the capture loads fine. In a previous iteration of my code where I didn't have these SYNC errors renderdoc did not crash. I decided to redesign the pipeline barriers slightly differently hence there being a difference.

Here is an exampe capture that causes the issue : 

[renderdoc-segfault-capture.rdc.gz](https://github.com/baldurk/renderdoc/files/6419794/renderdoc-segfault-capture.rdc.gz)

It is gzipped
It crashes when I try to load it in to renderdoc

## Environment

RenderDoc version: RenderDoc v1.14 Linux 64-bit Development (f530392)    (Built from source on linux from latest source as of May 4th)
Operating System: Ubuntu 21.04
Graphics API: Vulkan

Graphics card : GPU Intel 620 integrated

**Comment by baldurk** [comment]
This was caused by a mapped memory overflow when masking out mapped memory writes to memory that has tiled resources bound.

Although this is legal I strongly recommend that you separate memory allocations used for CPU-side mapping and then _do not_ map memory which has tiled resources bound to it on an integrated GPU like this. It has a performance penalty on the replay side to ensure that the tiled resources don't get overwritten on replay by detected map writes.

**Comment by bobtedbob** [comment]
Thanks as an F.Y.I. if I change the image to have tiling VK_IMAGE_TILING_LINEAR instead of VK_IMAGE_TILING_OPTIMAL renderdoc doesn't crash. But it is recommended to not use LINEAR.

I used a staging buffer for the data transfer (understand that it maybe slow but its something I will change later), I map it copy, unmap then destroy.

Can you elaborate more on the mapped memory overflow part of your statement? All image buffers are tiled (linear or optimal). Can you expand on what is overflowing?

**Comment by baldurk** [comment]
Yes I'm definitely not suggesting that you use linear tiling. That likely just works around the issue because it means that renderdoc doesn't mask the memory writes and so didn't go down the broken path.

The issue is that you have memory objects with optimal tiled resources which are mapped on the CPU. On a discrete GPU this wouldn't happen because your optimal tiled resources would be in VRAM and not able to be mapped. What I would recommend is that even on an integrated CPU that you treat any memory allocations that will have optimal tiled resources as if they were not CPU accessible and don't map them. Otherwise RenderDoc will detect writes from the GPU side during rendering, think they are CPU-side modifications, and try to record them into the capture. On replay I then need to detect when this is the case and omit it but it means that replaying mapped memory updates is no longer a simple memcpy.

The memory overflow was just me explaining the bug on RenderDoc's side. At one point there was a detected write to mapped memory from some offset with a certain size. When RenderDoc replayed the map it detected that some parts of the write overlapped with GPU optimal tiled resources so it skipped those, but when the last part of the write overlapped with linear memory it copied the whole bit of linear memory - even though only some of it was mapped/written to. This overflow was on RenderDoc's side, not yours.

The bug has been fixed which is why the github issue is closed. See above my previous comment for the link to the commit which fixes it.

**Comment by bobtedbob** [comment]
Thanks again for taking the time to explain things. I had two memory buffers in place, a staging one (CPU_ONLY) and a GPU one (GPU_ONLY) (using VMA speak). The GPU_ONLY one is the actual buffer used by the image (so TILED_OPTIMAL). I mapped, copied data and unmapped the CPU one and then used vkCmdCopyBufferToImage to copy the CPU_ONLY one to the GPU_ONLY one so I never mapped the GPU buffer directly. In my case the integrated gpu will still map the GPU_ONLY buffer in cpu memory (because its shared).

Confirmed the issue is fixed

**Comment by baldurk** [comment]
Whether the GPU internally has CPU visible virtual address space allocated is not visible outside of the driver, I'm only talking about vulkan maps of the memory from the application, and in the capture you shared the memory allocation that the image was bound to was mapped. If you're using VMA perhaps you're passing a flag or otherwise configuring it such that it maps the memory for you.

You can check if you're still hitting the slow path as there will be a warning added to the warnings & errors window (which will be indicated in the status bar on loading a capture) saying:

> Unmapped memory overlaps tiled-only memory region. Taking slow path to mask tiled memory writes.

**Comment by werman** [comment]
I have a similar crash in the same `memcpy`, however c7b7d3dbcc0c61dad3837e7c5abaaeac99c31880 doesn't fix it. From a quick look I think that `size = RDCMIN(MemRange.size, size);` should be `size = RDCMIN(MemRange.size - offs, size);` since otherwise memcpy could overrun `MappedData` by `offs` bytes (which fixes the crash).

The capture was made and replayed on Turnip driver (open source VK driver for Adreno). So I don't know how useful the capture would be to reproduce the crash.

**Comment by werman** [comment]
Here is the values at the moment of memcpy, if it doesn't make immediate sense I'll open a proper issue.

```
(gdb) p MemRange
$5 = {sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE, pNext = 0x0, memory = 0x55555df5a0, offset = 13164552, size = 53893077}
(gdb) p it->start()
$6 = 38768640
(gdb) p it->finish()
$7 = 67036928
(gdb) p size
$8 = 41432840
(gdb) p offs
$9 = 25604088
(gdb) p finish
$10 = 67057629

```

**Comment by baldurk** [comment]
I can see where there is a hole and it took me a while to convince myself of that, let alone if your fix was correct, so I the code needs to be refactored either way. It looks like the problem is in transforming to a map-relative offset too soon before that clamp. I've pushed bd3cf5c166c5b73265946b2bf61135c09d64b5b0 that instead clamps in absolute offsets for start/end of the map region and the tiled resource region, then rebase once the region has been decided. Indeed I would have no way of reproducing if the capture required adreno since I only have android devices that use adreno.

Everything I said above still applies though, I've fixed this but it's a slow path and it won't always work so I strongly recommend you don't map memory that will be used to back tiled resources. E.g. if you have aliased resources such that tiled and linear resources both sit in memory the writes will not be masked out because it's impossible to know whether it's GPU writes to the tiled resource that shouldn't be replayed, or CPU writes to the tiled resource which should be replayed. Doing the wrong thing can lead to a GPU crash writing into a tiled resource in use.
